File: Math/Java/2_Count_Primes.java
/*
 Question - Count Prime
 In order to solve the problem first let us understand what are prime numbers
 Prime numbers are the numbers which are divisible by 1 and by itself
 Step 1 : Taken value of N from user
 Step 2 : Called countPrimeNumbers Function to find no. of prime numbers
 Step 3 : Iterate from 2 to N and check which number is prime by calling isPrime Function which 
 returns false if any value is divisible from 2 to num/2 otherwise return true. If the value is 
 return by isPrime Function is true then we increase count 
 Step 4 : Return count from countPrimeNumbers Function and print it in the main function.
 Time Complexity : O(n^2)
 Space Complexity : O(1)
*/
import java.util.*;
public class Count_Primes
{
    //Function to check whether a number is prime or not
    public static boolean isPrime(int num) {
        
        for(int x = 2; x <= num/2; x++) 
        {
            if(num % x == 0) // Checking whether number is divisible by values in x
            {
                return false; // Returning false if remainder is 0
            }
        }
        return true; 
    }
    
    //Function to check number of Prime Numbers
    public static int countPrimeNumbers(int n) {
        
        int count = 0;//count number of prime numbers
        // Since 1 is neither prime nor composite , we will start counting from 2
        for(int i = 2; i <= n; i++) // Find number of prime numbers from 2 to n
        {
            if(isPrime(i)) //Check if a number is prime or not by calling is Prime Function
            {
                count++; // Increase count if number is prime
            }
        }
        
        return count; //Return value of count
    }
    
    public static void main(String args[]) {
        
        Scanner sc=new Scanner (System.in);
        int N = sc.nextInt(); // Taking User Input
         
        int result = countPrimeNumbers(N); // Calling countPrimeNumbers Function to find no. of prime numbers
        System.out.println(result);// Printing Result
    }
}

//Input 1 : 11     Output 1 : 5
//Input 2 : 1      Output 2 : 0

File: Math/Python/2_Count_Primes.py
# Count the number of prime numbers less than a non-negative number, n
# problem statement: https://leetcode.com/problems/count-primes/

"""
Solution: Simple solution, first find the prime numbers less than n and increment the counter when ever you find prime number.
"""
class Solution:
    def countPrimes(self, n: int) -> int:
        count = 0             # set counter to 0
        for i in range(2, n + 1):      # start  loop 2 to n
            if i >= 1:
                for j in range(2, (i // 2) + 1):      # loop for find prime
                    if i % j == 0:            # check number is divisible or not
                        break        # number is divisible then break
                else:
                    count += 1      # not divisible increment counter
        return count       # return counter value

num = int(input("enter Number:"))      # take input from user
print(Solution.countPrimes(Solution,num))    # calling to function

''' 
Example 1:
Input: n = 10
Output: 4
Explanation: There are 4 prime numbers less than 10, they are 2, 3, 5, 7.

Example 2:
Input: n = 25
Output: 9
Explanation: There are 9 prime numbers less than 25, they are 2, 3, 5, 7, 11, 13, 17, 19, 23.

Example 3:
Input: n = 50
Output: 15
Explanation: There are 15 prime numbers less than 50, they are 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47.

Example 4:
Input: n = 0
Output: 0
Explanation: There are no prime numbers less than 0.

Example 5:
Input: n = 1
Output: There are no prime numbers less than 1.
'''